File: Python/yamakawa/fsPCA.py
import numpy as np

def duplo(x):
    y = np.hstack((x, x**2))
    return y

class TfsPCA:
    def __init__(self, nVariaveis):
        n = 2 * nVariaveis
        self.St = np.zeros((1,n))
        self.St2 = np.zeros((n,n))
        self.C = np.zeros((n,n))
        self.constPCA = 0.90
        self.nVariaveis = nVariaveis

    def fsPCA(self, xt, ydt, h):
        xt = duplo(xt)
        n = np.shape(xt)[1]
        self.St = self.St + xt
        Mt = self.St/h # mean(xt,1)

        for i in range(0,n):
            for j in range(i,n):
                self.St2[i,j] = self.St2[i,j] + xt[:,i] * xt[:,j]
            for j in range(0,i):
                self.St2[i,j] = self.St2[j,i]
            
        cols = np.arange(0,self.nVariaveis)
        if h == 1:
            return cols
        Mt2 = self.St2/h
        # C = cov(xt)
        for i in range(0,n):
            for j in range(i,n):
                self.C[i,j] = Mt2[i,j] - Mt[:,i] * Mt[:,j]
            for j in range(0,i):
                self.C[i,j] = self.C[j,i]

        d, V = np.linalg.eig(self.C)
        d = d/sum(d)
        idx = np.argsort(d, 0)
        idx = idx[::-1]
        x = d[idx]
        m = np.min(np.where(np.cumsum(x) > self.constPCA))
        cols = idx[0:m]
        return cols

File: Python/yamakawa/temp2/fsPCA.m
function [m, cols, V] = fsPCA(nVariaveis, xt, ydt, constContribGlobal, h)
    constPCA = 0.90;
    global St;
    global St2;
    xt = duplo(xt);
    nVariaveis = size(xt, 2);
    St = St + xt;
    Mt = St/h; % mean(xt,1)

	for i = 1:nVariaveis
		for j = i:nVariaveis
			St2(i,j) = St2(i,j) + xt(:,i) * xt(:,j);
		end
		for j = 1:i-1
			St2(i,j) = St2(j,i);
		end
	end
        
    if h == 1
        m = nVariaveis;
        cols = 1:m;
        V = eye(nVariaveis);
        return;
    end
    Mt2 = St2/h;

	% C = cov(xt)
	for i = 1:nVariaveis
		for j = i:nVariaveis
			C(i,j) = Mt2(i,j) - Mt(i) * Mt(j);
		end
		for j = 1:i-1
			C(i,j) = C(j,i);
		end
	end
%     [V,D] = qdwheig(C,[],0);
%     d = diag(D);
    [V,d] = eig(C,'vector');
    d = d/sum(d);
    d = sort(d, 'desc');
    m = min(find(cumsum(d) >= constPCA));
    V = real(V(:,end-m+1:end));
    cols = 1:m;
end